### PR TITLE
Issue #922 Add checks when updating asset's backing asset id

### PR DESCRIPTION
**Pull request by jmjatlanta**
Issue #922 

This adds some of the same backing_asset checks to updates that are there during asset creation.

Note: This also fixes a bug found whereby some checks were examining the existing backing asset's properties instead of those of the new backing asset.

ToDo:
- [ ] add tests

### CODE REVIEW

**Comment by abitmore** [comment]
Good catch. Unfortunately this is a hard fork change. @ryanRfox, @pmconrad and @oxarbitrage: do you think we should put this fix into the coming hf release? If yes, we need a hotfix for testnet release.

File: libraries/chain/asset_evaluator.cpp [hunk 1]
```diff
@@ -360,42 +360,60 @@ void_result asset_update_issuer_evaluator::do_apply(const asset_update_issuer_op
...
    {
-      FC_ASSERT(a.dynamic_asset_data_id(d).current_supply == 0);
-      FC_ASSERT(d.find_object(o.new_options.short_backing_asset));
+      FC_ASSERT(asset_obj.dynamic_asset_data_id(d).current_supply == 0,
+            "Cannot update a bitasset if there is already a current supply.");
+      FC_ASSERT(d.find_object(op.new_options.short_backing_asset),
+            "The short backing asset could not be found");
 
-      if( a.issuer == GRAPHENE_COMMITTEE_ACCOUNT )
+      const asset_object& new_backing_asset = op.new_options.short_backing_asset(d);
```
Comment (abitmore) [inline libraries/chain/asset_evaluator.cpp]: Can combine this assignment with the `find_object` call above for better performance.

File: libraries/chain/asset_evaluator.cpp [hunk 2]
```diff
@@ -360,42 +360,60 @@ void_result asset_update_issuer_evaluator::do_apply(const asset_update_issuer_op
...
+               "The precision of the asset and backing asset must be equal." );
+      }
+
+      if( asset_obj.issuer == GRAPHENE_COMMITTEE_ACCOUNT )
+      {
+         if( new_backing_asset.is_market_issued() )
          {
-            const asset_object& backing_backing = backing.bitasset_data(d).options.short_backing_asset(d);
+            const asset_object& backing_backing = new_backing_asset.bitasset_data(d).options.short_backing_asset(d);
             FC_ASSERT( backing_backing.get_id() == asset_id_type(),
```
Comment (abitmore) [inline libraries/chain/asset_evaluator.cpp]: No need to fetch  `asset_object` then call `get_id()`. Use `short_backing_asset` field directly.

**Comment by pmconrad** [comment]
I think this is a bug that should be fixed, the sooner the better.
I'd say we fix this in current release branch, then create a new testnet release from current release branch. Perhaps postpone the testnet hf by a couple of days.

File: libraries/chain/asset_evaluator.cpp [hunk 3]
```diff
@@ -360,42 +360,60 @@ void_result asset_update_issuer_evaluator::do_apply(const asset_update_issuer_op
...
-                       "May not create a blockchain-controlled market asset which is not backed by CORE.");
+                       "May not modify a blockchain-controlled market asset to one which is not backed by CORE.");
          } else
-            FC_ASSERT( backing.get_id() == asset_id_type(),
-                       "May not create a blockchain-controlled market asset which is not backed by CORE.");
+            FC_ASSERT( new_backing_asset.get_id() == asset_id_type(),
+                       "May not modify a blockchain-controlled market asset to one which is not backed by CORE.");
```
Comment (abitmore) [inline libraries/chain/asset_evaluator.cpp]: "May not modify a blockchain-controlled market asset to one which is not backed by CORE." -> "May not modify a blockchain-controlled market asset to be backed by an asset which is not market issued asset nor CORE."

**Comment by ryanRfox** [comment]
Yes, I agre with @pmconrad to fix this now. Also, agree we should create a new TESTNET release with new hard fork date. Delay to perhaps MON 21 MAY. I don't feel any other dates need to be pushed as a result. We will need to coordinate with TESTNET block producers to ensure they take the new code.

**Comment by jmjatlanta** [comment]
Just to make sure I'm clear. Please correct me if I'm wrong:

1) I based this PR on the ``hardfork`` branch, as the additional checks belong to the next hardfork. That should not change.
2) The bug found is urgent, and the fix should be based on the ``release`` branch. The ``release`` branch changes will eventually make their way to the ``hardfork`` branch, which will probably cause a small conflict, but we'll sort it out at that time.

**Comment by jmjatlanta** [comment]
Rebased this branch from hardfork to release. Closing this PR and will open another.

**Pull request closed without merging**